src/Symfony/Component/ObjectMapper/Tests/ObjectMapperTest.php

```diff
@@ -0,0 +1,136 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\ObjectMapper\Tests;
+
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\ObjectMapper\Exception\RuntimeException;
+use Symfony\Component\ObjectMapper\ObjectMapper;
+use Symfony\Component\ObjectMapper\ReflectionMapperMetadataFactory;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\A;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\B;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\C;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\D;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\DeeperRecursion\Recursive;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\DeeperRecursion\RecursiveDto;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\DeeperRecursion\Relation;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\DeeperRecursion\RelationDto;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\InstanceCallback\A as InstanceCallbackA;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\InstanceCallback\B as InstanceCallbackB;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\MapStruct\AToBMapper;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\MapStruct\MapStructMapperMetadataFactory;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\MapStruct\Source;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\MapStruct\Target;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\MultipleTargets\A as MultipleTargetsA;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\MultipleTargets\C as MultipleTargetsC;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\Recursion\AB;
+use Symfony\Component\ObjectMapper\Tests\Fixtures\Recursion\Dto;
+use Symfony\Component\PropertyAccess\PropertyAccess;
+
+final class ObjectMapperTest extends TestCase
+{
+    /**
+     * @dataProvider mapProvider
+     */
+    public function testMap($expect, $args, array $deps = [])
+    {
+        $mapper = new ObjectMapper(...$deps);
+        $this->assertEquals($expect, $mapper->map(...$args));
+    }
+
+    /**
+     * @return array{expect: object, args: array, deps: array}
+     */
+    public function mapProvider()
+    {
+        $d = new D(baz: 'foo', bat: 'bar');
+        $c = new C(foo: 'foo', bar: 'bar');
+        $a = new A();
+        $a->foo = 'test';
+        $a->transform = 'test';
+        $a->baz = 'me';
+        $a->notinb = 'test';
+        $a->relation = $c;
+        $a->relationNotMapped = $d;
+
+        $b = new B('test');
+        $b->transform = 'TEST';
+        $b->baz = 'me';
+        $b->nomap = true;
+        $b->concat = 'testme';
+        $b->relation = $d;
+        $b->relationNotMapped = $d;
+        yield [$b, [$a]];
```

Comment:
I think that this is a bug due to how constructors are handled.

Reply:
Edit: can you give me a full example, not sure with all the fixtures I'd prefer an example from scratch

Reply:
@soyuka of courde, i just played a bit with your test data, changed class A:
```
#[Map(B::class)]
class A
{
    #[Map('bar')]
    public string $foo;

    public string $baz;

    public string $notinb;

    #[Map(transform: 'strtoupper')]
    public string $transform;

    #[Map(transform: [self::class, 'concatFn'])]
    public ?string $concat = null;

    #[Map(if: 'boolval')]
    public bool $nomap = false;

    public C $relation;

    public D $relationNotMapped;

//new
    #[Map('ds')]
    public array $cs = [];

    public function getConcat()
    {
        return 'should';
    }

    public static function concatFn($v, $object): string
    {
        return $v.$object->foo.$object->baz;
    }
}
```
and Class B:
```
class B
{
    public function __construct(private string $bar)
    {
    }
    public string $baz;
    public string $transform;
    public string $concat;
    public bool $nomap = true;
    public int $id;
    public D $relation;
    public D $relationNotMapped;
//new
    public array $ds;
}
```
and the change in the `mapProvider()` from my comment above.

Reply:
okay I understand, yes we don't handle collections yet, it makes things complex for no reason in this first iteration especially that we want to avoid heavy type-checking, and as array typing is not a thing in PHP yet I prefer to leave that to more complex implementations of the Mapper (a generated one for example [#see rfc mapper discussion](https://github.com/symfony/symfony/issues/54476)).
In my opinion, you'd just loop and use `map` on objects when you encounter a collection which is acceptable in user-land for now.

Reply:
Ok ... I just thought about a very simple usecase:
I have an API with a blog post endpoint and the post has Comments. And I now want to map this from my internal model to the DTOs I return on the API. Then I of course also want that the Comments are converted to the APICommentDTO. Of course in a first step there are ways to handle that differently.
Of course this brings a complexity. And yes, maybe it's good to have this first MVP, but I think collection/array handling is a common usecase.